substrate/client/network/src/discovery.rs

```diff
@@ -785,16 +784,23 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 							Ok(GetRecordOk::FoundRecord(r)) => {
 								debug!(
 									target: "sub-libp2p",
-									"Libp2p => Found record ({:?}) with value: {:?}",
+									"Libp2p => Found record ({:?}) with value: {:?} id {:?} stats {:?}",
 									r.record.key,
 									r.record.value,
+									id,
+									stats,
 								);
 
 								// Let's directly finish the query, as we are only interested in a
 								// quorum of 1.
 								if let Some(kad) = self.kademlia.as_mut() {
-									if let Some(mut query) = kad.query_mut(&id) {
-										query.finish();
+									if let Some(_query) = kad.query_mut(&id) {
+										// Let the query continue, to increase the chances we
+										// discover all possible addresses, for the cases where more
+										// addresses might exist in DHT, for example when the node
+										// changes its PeerId.
+
+										// query.finish();
```

Comment:
Should we still impose a limit on received results? Just to end the query a bit early, say 16 results?

Reply:
Probably makes sense to double check that our code works as expected when a record is returned more than once — e.g., the authority-discovery would have generated an error in this case before this PR.

Reply:
By the way, libp2p kad spec defines a mechanism for selecting the most up to date record among the returned (and even updating the outdated records):
https://github.com/libp2p/specs/blob/master/kad-dht/README.md#value-retrieval

But I don't think we can select the most recent PeerId at this point, can't we?

Reply:
> Should we still impose a limit on received results? Just to end the query a bit early, say 16 results?

That seems to be already bound by the replication factor which is 20.

> Probably makes sense to double check that our code works as expected when a record is returned more than once.

Any particular path that you think might be troublesome ?

Reply:
> > Probably makes sense to double check that our code works as expected when a record is returned more than once.
> 
> Any particular path that you think might be troublesome ?

Nothing specific, just thinking out loud.